ProcessMaker/processmaker
# Indexed Search

## Changes
- Add Laravel Scout support to improve fulltext (non-PMQL) searches within Task and Request Saved Searches
- Add `indexed-search:enable` and `indexed-search:disable` commands to Artisan
- Add ability for packages to support indexed search
  - Specifically, support for Collections Saved Searches is enabled by https://github.com/ProcessMaker/package-collections/pull/202
- Fixes an issue where pagination did not work properly when searching within a Saved Search
- Fixes an issue where the search needed to be submitted twice for it to work

## Requirements
- https://github.com/ProcessMaker/package-collections/pull/202
- https://github.com/ProcessMaker/package-savedsearch/pull/191

## Known Issues
- Searching will return a maximum of 500 records
  - This is a Laravel Scout limitation, but can be resolved simply by narrowing down the search further
- Using this search method is not as precise as using PMQL and may return more results than expected

## Testing Notes
- Ensure that the [Collections](https://github.com/ProcessMaker/package-collections/pull/202) and [Saved Search](https://github.com/ProcessMaker/package-savedsearch/pull/191) updates required to support this functionality are present in your packages
- Ensure your instance has over ten Requests, Tasks, and Collections records, respectively
  - A [sample dataset](https://github.com/ProcessMaker/processmaker/files/5065331/sample.csv.zip) can be imported to Collections
- Prior to enabling indexed search, navigate to a Saved Search for each data type (Requests, Tasks, Collections) and try searching for a term that exists within the dataset
  - On especially large datasets, this may take some time and may even time out
- Now, enable the indexed search by running `php artisan indexed-search:enable`
  - This may take quite a long time, especially on large datasets — this is normal
- Once again, navigate to a Saved Search for each data type (Requests, Tasks, Collections) and try searching for a term that exists within the dataset
  - Even on large datasets, the results should be returned fairly quickly
  - Ensure that results are visible and that pagination works as expected
- Now disable the indexed search by running `php artisan indexed-search:disable`
  - Ensure that a message displays, telling you that indexed search has been disabled
  - Ensure that you can still search within a Saved Search (although it may be slow again on large datasets)

Closes https://processmaker.atlassian.net/browse/FOUR-1725.
